### ISSUE TITLE
Failed to download and register model using s3 presigned v4 url

### ISSUE THREAD

**Issue by vishal-wiai** [pytorch/serve#669]
I'm using management API to register a model with a HTTPs URL. It is a s3 presigned url. Torchserve successfully downloads and registers models from s3 v2 signature URLs but fails to download for v4 signature types. Is there any bug to it?

I checked this behaviour with 3 of the models and also verified URL permissions. Using `requests.get()` I'm able to download but not on TorchServe.

**Comment by harshbafna** [comment]
@vishal-wiai: I was able to successfully register a model using an s3 pre-signed v4 URL.

**Note**: I had to replace `&` characters with its corresponding URL encoding character in the curl command i.e., `%26`.

**Comment by vishal-wiai** [comment]
Yes thank you @harshbafna . It worked after encoding `&` in the url with `%26`. 

As a feature request, is it possible for TorchServe to do this encoding, under the hood?

**Comment by vishal-wiai** [comment]
Is it always safe to replace all `&` with `%26` in the url as string manipulation or are there any cases where this will fail?

**Comment by vishal-wiai** [comment]
It works when you only provide the `url` parameter but fails when you provide additional params like `initial_workers`, `batch_size`, etc..
Any other encoding required?

**Comment by maaquib** [comment]
@vishal-wiai Can you provide the HTTPs URL alongwith the parameters you are using?

**Comment by harshbafna** [comment]
@vishal-wiai: 

> As a feature request, is it possible for TorchServe to do this encoding, under the hood?

This is the way an HTTP request works, where the parameters supplied in the URL using `?` are separated using  `&` character. Thus, if the data to be passed to the URL, in your case the model URL, contains any `&` character, it needs to be encoded using the corresponding hexadecimal value. Hence, this needs to be handled at the client end.









> It works when you only provide the `url` parameter but fails when you provide additional params like `initial_workers`, `batch_size`, etc..
> Any other encoding required?

You only need to encode the `&` character in the URL data only. The other parameters should be separated using `&` character only.

**Example**
```
curl -X POST 'http://localhost:8081/models?url=<http_encoded_presigned_s3_url>&initial_workers=1'
```

**Comment by vishal-wiai** [comment]
@harshbafna CURL did work for me but using `requests` gives me `HTTP 400` error (for both encoded and decoded url). Am I missing something?

```
import requests

params = (('url', '<http-presign-s3-url>'), ('initial_workers', '1'))
response = requests.post('http://localhost:8081/models', params=params)
```

**Comment by harshbafna** [comment]
@vishal-wiai: This seems to be a problem with the `requests` module in python which by default encodes every special character supplied in the params.

A simple approach would be URL formation through string concatenation:

```
url = 'http://localhost:8081/models?url='+model_url.replace('&', '%26')+'&initial_workers=1'
response = requests.post(url)
```

Reference: https://github.com/psf/requests/issues/1454#issuecomment-20832874

**Comment by dotel-saramsz** [comment]
I tried all of the techniques suggested above but still could not get torchserve to download the model from s3 presigned url.
The presigned url I got is of the following format:

`https://<bucket-name>.s3.amazonaws.com/<object-name>?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=<some-credential>%2F<date_string>%2F<aws-region>%2F...`

When I send this url as part of the POST request to register the model:

`curl -X POST 'http://localhost:8081?url=<url_given_above>&model_name=mymodel`

I get back a `400 Bad Request` response indicating that the request failed due to bad syntax. I tried replacing the `&` in the presigned url with `%26` but to no avail.
The model can be downloaded from the presigned url through my browser and curl in my local machine but not through torchserve. Looks like if I have any query param (here, the AWS specific credentials) in the `url` param of the `models/` POST request, the request fails.

Any pointers would be appreciated.
